contracts/zero-ex/contracts/src/features/ERC721OrdersFeature.sol

```diff
@@ -0,0 +1,1041 @@
+            taker,
+            erc721TokenId
+        );
+
+        // Mark the order as filled.
+        _setOrderStatusBit(order.maker, order.nonce);
+
+        // Transfer the ERC721 asset to the buyer.
+        // If this function is called from the 
+        // `onERC721Received` callback the Exchange Proxy
+        // holds the asset. Otherwise, transfer it from 
+        // the seller.
+        _transferERC721AssetFrom(
+            order.erc721Token,
+            isCallback ? address(this) : taker,
+            order.maker,
+            erc721TokenId
+        );
+
+        if (unwrapNativeToken) {
+            // The ERC20 token must be WETH for it to be unwrapped.
+            rrequire(
+                order.erc20Token == WETH,
+                LibERC721OrdersRichErrors.ERC20TokenMismatchError(
+                    address(order.erc20Token),
+                    address(WETH)
+                )
+            );
+            // Transfer the WETH from the maker to the Exchange Proxy
+            // so we can unwrap it before sending it to the seller.
+            // TODO: Probably safe to just use WETH.transferFrom for some
+            //       small gas savings
+            _transferERC20TokensFrom(
+                WETH,
+                order.maker,
+                address(this),
+                order.erc20TokenAmount
+            );
+            // Unwrap WETH into ETH.
+            WETH.withdraw(order.erc20TokenAmount);
+            // Send ETH to the seller.
+            _transferEth(payable(taker), order.erc20TokenAmount);
+        } else {
+            // Transfer the ERC20 token from the buyer to the seller.
+            _transferERC20TokensFrom(
+                order.erc20Token, 
+                order.maker, 
+                taker, 
+                order.erc20TokenAmount
+            );
+        }
+
+        // The buyer pays the order fees.
+        _payFees(order, order.maker, false);
+
+        emit ERC721OrderFilled(
+            order.direction,
+            order.erc20Token,
+            order.erc20TokenAmount,
+            order.erc721Token,
+            erc721TokenId,
+            order.maker,
+            taker,
+            order.nonce
+        );
+    }
+
+    // Core settlement logic for buying an ERC721 asset.
+    // Used by `buyERC721` and `batchBuyERC721s`.
+    function _buyERC721(
+        LibERC721Order.ERC721Order memory order,
+        LibSignature.Signature memory signature,
+        uint256 ethAvailable
+    )
+        public
+        returns (uint256 ethSpent)
+    {
+        // Check that the order can be filled.
+        _validateSellOrder(order, signature);
+
+        // Mark the order as filled.
+        _setOrderStatusBit(order.maker, order.nonce);
+
+        // Transfer the ERC721 asset to the buyer (`msg.sender`).
+        _transferERC721AssetFrom(
+            order.erc721Token,
+            order.maker,
+            msg.sender,
+            order.erc721TokenId
+        );
+
```

Comment:
Thinking about throwing in a callback to `msg.sender` here (and in `_sellERC721` too), so we have a flow similar to UniV3:
```
maker -> taker transfer
taker callback
taker -> maker transfer
```
This would enable matching 0x orders with e.g. Wyvern orders without a flash loan
Would make the `ethAvailable` accounting a bit more complex though but I think we can just do something like this:
```
uint256 balanceBeforeCallback = address(this).balance;
// taker callback
ethAvailable += address(this).balance - balanceBeforeCallback;
```

Reply:
(I did this too)